spec/lib/site_cache_spec.rb

```diff
@@ -0,0 +1,261 @@
+
+  def initialize(global_limit, site_limit, path)
+    @path = path
+    @cache = MethodLogger.new(SiteCache.new(global_limit, site_limit))
+    @keys = Hash.new { |h, k| h[k] = [] }
+  end
+
+  def generate_lookup_for_site(site_id)
+    operation = path.choose(:lookup_existing, :delete_existing)
+
+    public_send(operation, site_id)
+    @cache.operations.last
+  end
+
+  def generate_lookup
+    operation = path.choose(:lookup_existing, :delete_existing)
+
+    public_send(operation)
+    @cache.operations.last
+  end
+
+  def generate_operation
+    operation =
+      path.choose(
+        :insert_new,
+        :insert_existing_site_id,
+        :insert_existing_key,
+        :lookup_existing,
+        :getset_existing,
+        :delete_existing,
+        :clear,
+        :clear_existing_site,
+      )
+
+    public_send(operation)
+    @cache.operations.last
+  end
+
+  def generate_operations(upto:)
+    CacheOperations.new(path.choose(*(1..upto)).times.map { generate_operation })
+  end
+
+  def insert_new
+    site_id = Object.new
+    key = Object.new
+    keys[site_id] << key
+    value = Object.new
+
+    if path.choose(true, false)
+      cache.set(site_id, key, value)
+    else
+      cache.getset(site_id, key) { value }
+    end
+    [site_id, key, value]
+  end
+
+  def insert_existing_site_id
+    site_id = path.choose(*keys.keys)
+    key = Object.new
+    keys[site_id] << key
+    value = Object.new
+
+    if path.choose(true, false)
+      cache.set(site_id, key, value)
+    else
+      cache.getset(site_id, key) { value }
+    end
+  end
+
+  def insert_existing_key
+    site_id = path.choose(*keys.keys)
+    key = path.choose(*keys[site_id])
+    value = Object.new
+    cache.set(site_id, key, value)
+  end
+
+  def lookup_existing(site_id = nil)
+    site_id ||= path.choose(*keys.keys)
+    key = path.choose(*keys[site_id])
+    cache.lookup(site_id, key)
+  end
+
+  def getset_existing
+    site_id = path.choose(*keys.keys)
+    key = path.choose(*keys[site_id])
+    value = Object.new
+    cache.getset(site_id, key) { value }
+  end
+
+  def delete_existing(site_id = nil)
+    site_id ||= path.choose(*keys.keys)
+    key = path.choose(*keys[site_id])
+    cache.delete(site_id, key)
+  end
+
+  def clear
+    cache.clear
+  end
+
+  def clear_site(site_id)
+    cache.clear_site(site_id)
+  end
+
+  def clear_existing_site
+    site_id = path.choose(*keys.keys)
+    cache.clear_site(site_id)
+  end
+end
+
+class CacheOperations
+  def initialize(operations)
+    @operations = operations
+  end
+
+  def values_inserted_at(site_id, key)
+    @operations.filter_map do |op|
+      case op.method_name
+      when :set
+        op_site_id, op_key, op_value = op.args
+        op_value if site_id == op_site_id && key == op_key
+      when :getset
+        op_site_id, op_key = op.args
+        op_value = op.blk.call
+        op_value if site_id == op_site_id && key == op_key && op_value == op.result
+      end
+    end
+  end
+end
```

Comment:
I personally find these additional classes very confusing and I'm wondering why we can't test the cache using the public methods of the `SiteCache` class instead introducing an additional layer of indirection.

Reply:
This is doing exhaustive testing.